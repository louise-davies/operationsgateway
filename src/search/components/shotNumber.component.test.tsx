import React from 'react';
import ShotNumber, { type ShotNumberProps } from './shotNumber.component';
import {
  screen,
  render,
  type RenderResult,
  within,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';

describe('shotNumber search', () => {
  let props: ShotNumberProps;
  const changeSearchParameterShotnumMin = jest.fn();
  const changeSearchParameterShotnumMax = jest.fn();
  const resetDateRange = jest.fn();
  let user;

  const createView = (): RenderResult => {
    return render(<ShotNumber {...props} />);
  };

  beforeEach(() => {
    props = {
      changeSearchParameterShotnumMin,
      changeSearchParameterShotnumMax,
      resetDateRange,
      isDateToShotnum: false,
    };

    user = userEvent.setup();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly', async () => {
    const { asFragment } = createView();
    await user.click(screen.getByLabelText('open shot number search box'));
    expect(asFragment()).toMatchSnapshot();
  });

  it('can open and close its popup window', async () => {
    createView();

    await user.click(screen.getByLabelText('open shot number search box'));
    const shotnumPopup = screen.getByRole('dialog');
    expect(
      within(shotnumPopup).getByText('Select your shot number')
    ).toBeInTheDocument();
    await user.click(screen.getByLabelText('close shot number search box'));
    expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
    expect(
      screen.queryByText('Select your shot number')
    ).not.toBeInTheDocument();
  });

  it('allows user to change min and max values and not show error when valid input', async () => {
    createView();

    await user.click(screen.getByLabelText('open shot number search box'));
    const shotnumPopup = screen.getByRole('dialog');
    const minInput = within(shotnumPopup).getByRole('spinbutton', {
      name: 'Min',
    });
    const maxInput = within(shotnumPopup).getByRole('spinbutton', {
      name: 'Max',
    });

    await user.type(minInput, '1');
    await user.type(maxInput, '2');
    expect(changeSearchParameterShotnumMin).toHaveBeenCalledWith(1);
    expect(changeSearchParameterShotnumMax).toHaveBeenCalledWith(2);
    expect(resetDateRange).toHaveBeenCalled();
    const helperTexts = within(shotnumPopup).queryAllByText('Invalid range');
    expect(helperTexts.length).toEqual(0);
  });

  it('displays invalid range message when min > max', async () => {
    props = {
      ...props,
      searchParameterShotnumMin: 1,
      searchParameterShotnumMax: 0,
    };
    createView();

    await user.click(screen.getByLabelText('open shot number search box'));
    const shotnumPopup = screen.getByRole('dialog');
    const minInput = within(shotnumPopup).getByRole('spinbutton', {
      name: 'Min',
    });
    const maxInput = within(shotnumPopup).getByRole('spinbutton', {
      name: 'Max',
    });

    expect(minInput).toHaveValue(1);
    expect(maxInput).toHaveValue(0);

    const helperTexts = within(shotnumPopup).getAllByText('Invalid range');
    // One helper text below each input
    expect(helperTexts.length).toEqual(2);
  });

  describe('displays the currently selected shot number range', () => {
    it('none', () => {
      createView();
      expect(screen.getByText('Select')).toBeInTheDocument();
    });

    it('minimum only', () => {
      props = {
        ...props,
        searchParameterShotnumMin: 1,
      };
      createView();
      expect(screen.getByText('Minimum: 1')).toBeInTheDocument();
    });

    it('maximum only', () => {
      props = {
        ...props,
        searchParameterShotnumMax: 1,
      };
      createView();
      expect(screen.getByText('Maximum: 1')).toBeInTheDocument();
    });

    it('minimum and maximum', () => {
      props = {
        ...props,
        searchParameterShotnumMin: 1,
        searchParameterShotnumMax: 2,
      };
      createView();
      expect(screen.getByText('1 to 2')).toBeInTheDocument();
    });
  });
});
