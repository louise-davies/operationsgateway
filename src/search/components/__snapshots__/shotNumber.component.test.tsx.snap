// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shotNumber search renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-79elbk"
  >
    <div
      aria-label="close shot number search box"
      class="MuiBox-root css-o3ftni"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hspsbx-MuiSvgIcon-root"
        data-testid="AdjustIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 2C6.49 2 2 6.49 2 12s4.49 10 10 10 10-4.49 10-10S17.51 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm3-8c0 1.66-1.34 3-3 3s-3-1.34-3-3 1.34-3 3-3 3 1.34 3 3z"
        />
      </svg>
      <div>
        <p
          class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap css-fvtpii-MuiTypography-root"
        >
          Shot Number
        </p>
        <h6
          class="MuiTypography-root MuiTypography-subtitle1 MuiTypography-noWrap css-b87fj0-MuiTypography-root"
        >
          Select
        </h6>
      </div>
    </div>
    <div
      class="MuiBox-root css-1997yk"
      role="dialog"
    >
      <div
        class="MuiBox-root css-ha9sb9"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-um3j9i-MuiTypography-root"
        >
          Select your shot number
        </p>
        <hr
          class="MuiDivider-root MuiDivider-fullWidth css-vy62gc-MuiDivider-root"
        />
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 css-a5rdam-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5 css-k3f0g4-MuiGrid-root"
          >
            <div
              class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for=":r0:"
                id=":r0:-label"
              >
                Min
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
                  id=":r0:"
                  min="0"
                  name="shot number min"
                  type="number"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-yjsfm1"
                  >
                    <span>
                      Min
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-u78q51-MuiGrid-root"
          >
            <p>
              to
            </p>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5 css-k3f0g4-MuiGrid-root"
          >
            <div
              class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for=":r1:"
                id=":r1:-label"
              >
                Max
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
                  id=":r1:"
                  min="0"
                  name="shot number max"
                  type="number"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-yjsfm1"
                  >
                    <span>
                      Max
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
